http-common/src/connector.rs

```diff
@@ -51,56 +57,108 @@ impl Incoming {
+                let accept = listener.accept();
+                futures_util::pin_mut!(accept);
+
+                match future::select(shutdown_loop, accept).await {
+                    future::Either::Left((_, _)) => break,
+                    future::Either::Right((tcp_stream, shutdown)) => {
+                        let tcp_stream = tcp_stream?.0;
+
+                        let server = crate::uid::UidService::new(None, 0, server.clone());
+
+                        tasks.fetch_add(1, std::sync::atomic::Ordering::AcqRel);
+                        let server_tasks = tasks.clone();
+                        tokio::spawn(async move {
+                            if let Err(http_err) = hyper::server::conn::Http::new()
+                                .serve_connection(tcp_stream, server)
+                                .await
+                            {
+                                log::info!("Error while serving HTTP connection: {}", http_err);
+                            }
+
+                            server_tasks.fetch_sub(1, std::sync::atomic::Ordering::AcqRel);
+                        });
+
+                        shutdown_loop = shutdown;
                     }
-                });
+                }
             },
 
             Incoming::Unix {
                 listener,
                 user_state,
             } => loop {
-                let (unix_stream, _) = listener.accept().await?;
-                let ucred = unix_stream.peer_cred()?;
-                let user_state = user_state
-                    .entry(ucred.uid())
-                    .or_insert_with(|| {
-                        std::sync::Arc::new(tokio::sync::Semaphore::new(MAX_REQUESTS_PER_USER))
-                    })
-                    .clone();
-
-                let server = crate::uid::UidService::new(ucred.pid(), ucred.uid(), server.clone());
-                tokio::spawn(async move {
-                    match user_state.try_acquire_owned() {
-                        Ok(_permit) => {
-                            if let Err(http_err) = hyper::server::conn::Http::new()
-                                .serve_connection(unix_stream, server)
-                                .await
-                            {
-                                log::info!("Error while serving HTTP connection: {}", http_err);
+                let accept = listener.accept();
+                futures_util::pin_mut!(accept);
+
+                // Await either the next established connection or the shutdown signal.
+                match future::select(shutdown_loop, accept).await {
+                    future::Either::Left((_, _)) => break,
+                    future::Either::Right((unix_stream, shutdown)) => {
+                        let unix_stream = unix_stream?.0;
+
+                        let ucred = unix_stream.peer_cred()?;
+                        let user_state = user_state
```

Comment:
Since you've "changed" this code anyway (by indenting it), can I convince you to get rid of this semaphore and replace it with another `AtomicUsize` ?

Reply:
Ok, did that.